### ISSUE TITLE
Outreachy project proposals (Oct. 2021)

### ISSUE THREAD

**Issue by joshmoore** [zarr-developers/community#39]
Zarr has been accepted as an [Outreachy project](https://www.outreachy.org/communities/cfp/zarr/) for mentoring interns during a three month paid project. Projects are propose by the mentors and should balance both the attractiveness and feasibility of the project along with the learning effect for the interns. Projects must be submitted by **September 29th**. Comments and ideas welcome (though complete drafts are more likely to be used :wink:).

### Contribution section:
During the application period, interns make contributions to open-source projects to give mentors a chance to evaluate their skills & interest. This same block of text will likely be used for all proposals under the "Contribution" section:

> Contributions to any of the repositories in the [zarr-developers](https://github.com/zarr-developers) organization are welcome. Particularly of interest are the ["help wanted"](https://github.com/search?q=org%3Azarr-developers+label%3A%22help+wanted%22&type=Issues) and ["low-hanging-fruit"](https://github.com/search?q=org%3Azarr-developers+label%3A%22low-hanging-fruit%22&type=Issues) labels. Most of the issues will be on the [zarr-python](https://github.com/zarr-developers/zarr-python) repo. If you are interested in getting involved in another language, see the list under ["Zarr implementations"](https://github.com/zarr-developers/zarr_implementations#zarr-implementations) and let us know what you are interested in under https://github.com/zarr-developers/community/discussions.

----

### Proposal 1. Add multi-language Zarr implementation tests (C/Julia/etc.)

[Zarr](https://zarr.readthedocs.io/en/stable/) is a chunked, compressed format for the storage of N-dimensional arrays like those created by [NumPy](https://numpy.org/). There are a number of Zarr implementations across a wide range of programming languages which all need to conform to the same specification. We would like to have Zarr tested across as many of these languages as possible in order to ensure conformance. Several languages are already well-represented in the [Zarr test suite](https://github.com/zarr-developers/zarr_implementations); others have not yet been added like include Julia ([\#42](https://github.com/zarr-developers/zarr_implementations/issues/42)) or C (netcdf [\#35](https://github.com/zarr-developers/zarr_implementations/issues/35); or tensorstore from Google [\#20](https://github.com/zarr-developers/zarr_implementations/issues/20)). Candidates should pick implementations in their order of preference and test them through several phases as described under “tasks”.

**Tasks incl. stretch goals:**

- Phase 1: A small driver application will be needed in each language. Initially, the driver should create a Zarr file using the given library and evaluate if the output can be read by other implementations.
- Phase 2: Extend the driver to reading Zarr files from all other implementations ([\#25](https://github.com/zarr-developers/zarr_implementations/issues/25)), reporting (and perhaps fixing!) any failures.
- Phase 3 (stretch): Add any test functions that may be of particular interest to a given implementation, reporting (and perhaps fixing!) any failures.

**Benefits:**

This project will give the prospective intern the opportunity to try out several new languages while learning the basics of the Zarr format.

**Community Benefits:**

As the Zarr ecosystem grows, testing that all the implementations produce compatible data is critical for maintaining user trust.

----

### Proposal 2. Build registry of Zarr codecs

[Zarr](https://zarr.readthedocs.io/en/stable/) is a chunked, compressed format for the storage of N-dimensional arrays like those created by [NumPy](https://numpy.org/). Each compression algorithm used by Zarr is assigned an identifier by the [numcodecs](https://github.com/zarr-developers/numcodecs) library. In order to make these identifiers useful in other programming languages, [issue \#278](https://github.com/zarr-developers/numcodecs/issues/278) requires a registry to be defined that is machine readable from multiple programming languages (C, Java, Javascript, etc.).

**Tasks incl. stretch goals:**

- Research and compare existing lists of compression filters (e.g. from [HDF5](https://support.hdfgroup.org/services/filters.html) and [imagecodecs](https://github.com/cgohlke/imagecodecs)).
- Document all of the current codecs and their identifiers from the research in an issue.
- Propose a format for the registry in an issue for community review.
- Create and publish the registry, both in a machine readable as well as human-friendly form.
- Document the steps for others to update the registry.
- Stretch goal: Write tests that show which codecs from each implementation are supported by other implementations. Display a matrix of the results.

**Benefits:**

This project will give the prospective intern insight into low-level details of a number of file formats and compression techniques as well as the importance of standardization for keeping data readable.

**Community Benefits:**

The registry will serve as a critical bridge between a number of communities (different programming languages like C/Java/Python as well as multiple file formats like Zarr/NetCDF/HDF5).

----

### Proposal 3. Benchmark Zarr implementations

[Zarr](https://zarr.readthedocs.io/en/stable/) is a chunked, compressed format for the storage of N-dimensional arrays like those created by [NumPy](https://numpy.org/). Performance when reading and writing very large arrays is critical for user acceptance. With [issue\#337](https://github.com/zarr-developers/zarr-python/issues/337), we would like to have benchmarks updated with each version of Zarr showing the long-term tendency of various metrics, warning developers if a change causes too much slowdown. [Airspeed Velocity (asv)](https://github.com/airspeed-velocity/asv) is a likely tool for this project. (Example output can be seen in the [aicsimageio](https://allencellmodeling.github.io/aicsimageio/_benchmarks/index.html) project.)

**Tasks incl. stretch goals:**

- Define a number of simple metrics for reading and writing Zarr files, asking the community for further suggestions. Examples include: writing arrays in a number of different sizes & shapes; reading those same arrays back in.
- Implement these metrics in zarr-python and optionally in other programming languages available in [zarr\_implementations](https://github.com/zarr-developers/zarr_implementations).
- A further stretch goal would be to compare the same operations using HDF5 or NumPy files (.npy) as described in [issue \#519](https://github.com/zarr-developers/zarr-python/issues/519).

**Benefits:**

This project will give the prospective intern the opportunity to learn the basics of the Zarr format and APIs as well as experiment with benchmarking, a useful tool for any programmer.

**Community Benefits:**

The community will benefit from a clear, visual representation of Zarr’s speed overtime, which will identify areas for further optimization**.**

----

### Proposal 4. Fuzz-test Zarr implementations

[Zarr](https://zarr.readthedocs.io/en/stable/) is a chunked, compressed format for the storage of N-dimensional arrays like those created by [NumPy](https://numpy.org/). Implementing libraries must reliably read the same data across a large number of parameters: a range of platforms and codecs, big and little endian systems, etc. [Fuzz testing](https://en.wikipedia.org/wiki/Fuzzing), “an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program”, will help to show that each library can cope with this diversity.

**Tasks incl. stretch goals:**

- Chose a first Zarr library for testing together with the mentors (e.g. [zarr-python](https://github.com/zarr-developers/zarr-python))
- Evaluate fuzz testing frameworks for the appropriate language and report back on a public issue.
- Choose a framework and automate fuzzing of the Zarr library with GitHub Actions.
- Optionally, choose a next library and repeat.

**Benefits:**

This project will give the prospective intern the opportunity to learn the basics of the Zarr format and APIs as well as experiment with fuzz testing, a useful tool for any programmer.

**Community Benefits:**

The community will benefit from the improved security of the Zarr libraries.

----

### Remaining possible projects:

These were ideas found while going through open issues and PRs. Thoughts on how to refine one or more of them are welcome.

* Too small / lower value ?
  - [Run a user survey](https://github.com/zarr-developers/community/issues/32)
  - Improve release with towncrier or similar
  - Update examples in docs
  - https://github.com/zarr-developers/zarr-developers.github.io updates
  - [Automate updates to contributors list](https://github.com/zarr-developers/zarr-python/issues/804)
  - [Add jupyter tests](https://github.com/zarr-developers/governance/issues/8) ([+tutorial](https://github.com/zarr-developers/zarr-python/issues/514))
  - Generate nice webpage of test results
* Too esoteric?
  - [Standardizing .zmetadata](https://github.com/zarr-developers/zarr-specs/issues/113)
  - [Awkward array](https://github.com/zarr-developers/community/issues/29)
  - [Transactions](https://github.com/zarr-developers/zarr-python/issues/247)
  - [tree bottleneck](https://github.com/zarr-developers/zarr-python/issues/224)
  - data apis: ([\_\_dask\_tokenize\_\_](https://github.com/zarr-developers/zarr-python/issues/202) or [ufuncs](https://github.com/zarr-developers/zarr-python/issues/566))
  - [find/findall](https://github.com/zarr-developers/zarr-python/issues/188),
  - [z-order](https://github.com/zarr-developers/zarr-python/issues/40)/[space-filling](https://github.com/zarr-developers/zarr-python/issues/90)
* Too big?
  - **[DataFrames](https://github.com/zarr-developers/zarr-python/issues/149) / ([Parquet](https://github.com/zarr-developers/zarr-python/issues/515))**
  - **[Store refactor](https://github.com/zarr-developers/zarr-python/issues/764) (also [414)](https://github.com/zarr-developers/zarr-python/issues/414)**
  - sparse ([CSR](https://github.com/zarr-developers/zarr-python/issues/723), [\#424](https://github.com/zarr-developers/zarr-python/issues/424), upstream: [pydata/sparse](https://github.com/pydata/sparse/issues/433))
  - [read h5](https://github.com/zarr-developers/zarr-python/issues/535) (in addition to fsspec-reference-maker)
  - [smart chunk access](https://github.com/zarr-developers/zarr-python/issues/547) / [memory
 mapping?](https://github.com/zarr-developers/zarr-python/issues/321)
  - [symlinks](https://github.com/zarr-developers/zarr-python/issues/298) (+[groups](https://github.com/zarr-developers/zarr-python/issues/297))
* Not fully defined:
  - Other implementations fsspec-reference-maker (Java, etc.) - needs jfsspec
  - Build an extension of the V3 spec (which one?)

----

Original drafted in [gdoc](https://docs.google.com/document/d/1EQ-tGpPehcqa6yfdvKUL-_avRTvrUKYmrTDk6nKAAhQ/edit#heading=h.13jm4cqgdlqg)

**Comment by joshmoore** [comment]
thanks for the :tada:s. I've now uploaded the proposals. I see that others from the community can also open their own and offer to co-mentor. If anyone is interested, get in touch.

**Comment by joshmoore** [comment]
In the end, we did not accept any of the potential Outreachy mentorees this round. It was lovely to interact with everyone on Zoom, Gitter, and GitHub, but ultimately I made the mistake of not making clear what would be needed in terms of existing Python experience in order to get involved with Zarr. My apologies for this! I will try to be clearer in the future. In that sense it was a good learning experience for _me_ and I thank everyone who participated.

The above outline is probably still useful for future proposals but I'll close this ticket for now.
